group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/CurrentAssignmentBuilder.java

```diff
@@ -0,0 +1,447 @@
+                // its state towards the latest target assignment.
+                return computeNextAssignment(
+                    member.memberEpoch() + 1,
+                    member.assignedTasks()
+                );
+
+            case UNRELEASED_TASKS:
+                // When the member is in the UNRELEASED_TASKS, we reconcile the
+                // member towards the latest target assignment. This will assign any
+                // of the unreleased tasks when they become available.
+                return computeNextAssignment(
+                    member.memberEpoch(),
+                    member.assignedTasks()
+                );
+
+            case UNKNOWN:
+                // We could only end up in this state if a new state is added in the
+                // future and the group coordinator is downgraded. In this case, the
+                // best option is to fence the member to force it to rejoin the group
+                // without any tasks and to reconcile it again from scratch.
+                if ((ownedTasks.isEmpty() || !ownedTasks.get().isEmpty())) {
+                    throw new FencedMemberEpochException(
+                        "The streams group member is in a unknown state. "
+                            + "The member must abandon all its tasks and rejoin.");
+                }
+
+                return computeNextAssignment(
+                    targetAssignmentEpoch,
+                    member.assignedTasks()
+                );
+        }
+
+        return member;
+    }
+
+    /**
+     * Takes the current currentAssignment and the targetAssignment, and generates three
+     * collections:
+     *
+     * - the resultAssignedTasks: the tasks that are assigned in both the current and target
+     * assignments.
+     * - the resultTasksPendingRevocation: the tasks that are assigned in the current
+     * assignment but not in the target assignment.
+     * - the resultTasksPendingAssignment: the tasks that are assigned in the target assignment but
+     * not in the current assignment, and can be assigned currently (i.e., they are not owned by
+     * another member, as defined by the `isUnreleasedTask` predicate).
+     */
+    private boolean computeAssignmentDifference(
+        Map<String, Set<Integer>> currentAssignment,
+        Map<String, Set<Integer>> targetAssignment,
+        Map<String, Set<Integer>> resultAssignedTasks,
+        Map<String, Set<Integer>> resultTasksPendingRevocation,
+        Map<String, Set<Integer>> resultTasksPendingAssignment,
+        BiPredicate<String, Integer> isUnreleasedTask
+    ) {
+        boolean hasUnreleasedTasks = false;
+
+        Set<String> allSubtopologyIds = new HashSet<>(targetAssignment.keySet());
+        allSubtopologyIds.addAll(currentAssignment.keySet());
+
+        for (String subtopologyId : allSubtopologyIds) {
+            Set<Integer> currentPartitions = currentAssignment.getOrDefault(subtopologyId, Collections.emptySet());
```

Comment:
Could you please rename the variables to use `Tasks` instead of `Partitions`? With `Partitions` it is a bit confusing.